src/IO/writers.jl

```diff
@@ -0,0 +1,391 @@
+abstract type AbstractWriter end
+
+
+using HDF5
+using ..Domains
+using ..Meshes
+using ..Topologies
+using ..Spaces
+using ..Fields
+
+"""
+    HDF5Writer
+
+A struct containing the `filename` and `cache` for writing data to the HDF5 file.
+The `cache` enables writing of relevant data without duplication.
+"""
+struct HDF5Writer <: AbstractWriter
+    file::HDF5.File
+    cache::Dict{String, String}
+end
+
+function HDF5Writer(filename::String)
+    file = h5open(filename, "w")
+    cache = Dict{String, String}()
+    return HDF5Writer(file, cache)
+end
+
+function cartesianindices_to_matrix(elemorder)
+    m, n = length(elemorder), length(eltype(elemorder))
+    elemordermatrix = zeros(Int, m, n)
+    for (i, order) in enumerate(elemorder)
+        for j in 1:n
+            elemordermatrix[i, j] = order[j]
+        end
+    end
+    return elemordermatrix
+end
+
+# write_new!(writer, obj) should return the name of the written object
+# write(writer, obj) checks the cache: if the object has not yet been written it then calls write_new!
+
+"""
+    write!(reader, obj[, path])
+
+Checks the cache to verify if the object, specified by the `path`, is in the cache.
+If it has not been read, this function reads it to the cache and returns the name of the object.
+"""
+write!(writer::HDF5Writer, obj, name = defaultname(obj)) =
+    get!(writer.cache, name) do
+        write_new!(writer, obj, name)
+    end
+
+# Domains
+defaultname(::Domains.SphereDomain) = "domains/sphere"
+defaultname(::Domains.IntervalDomain) = "domains/interval"
+
+"""
+    write_new!(writer, domain, name)
+
+Writes an object of type 'IntervalDomain' and name 'name' to the HDF5 file.
+"""
+function write_new!(
+    writer::HDF5Writer,
+    domain::Domains.IntervalDomain,
+    name::AbstractString = defaultname(domain),
+)
+    group = create_group(writer.file, name)
+    write_attribute(group, "type", "IntervalDomain")
+    write_attribute(group, "coord_type", string(typeof(domain.coord_min)))
+    write_attribute(group, "coord_min", string(domain.coord_min))
+    write_attribute(group, "coord_max", string(domain.coord_max))
+    !isnothing(domain.boundary_names) &&
+        write_attribute(group, "boundary_names", string(domain.boundary_names))
```

Comment:
write this as a collection of strings, not a single string
```suggestion
        write_attribute(group, "boundary_names", map(string, domain.boundary_names)))
```